src/main/java/frc/robot/subsystems/Vision.java

```diff
@@ -84,18 +87,90 @@ public PhotonTrackedTarget getBestTarget() {
      */
     public boolean isLinedUp() {
 
-        // Get the best target
-        PhotonTrackedTarget bestTarget = getBestTarget();
+        // Check to see if the yaw deviation is below the tolerance
+        return (Parameters.vision.YAW_TOLERANCE > getBestTarget().getYaw());
+    }
 
-        // Make sure that the target isn't null (that we have a target)
-        if (bestTarget != null) {
+    // return list of list for parseTargetCorners
+    private List<List<TargetCorner>> getTargetCorners() {
+        PhotonPipelineResult pipelineResult = camera.getLatestResult();
+        if (!pipelineResult.hasTargets()) return null;
 
-            // Check to see if the yaw deviation is below the tolerance
-            return (Parameters.vision.YAW_TOLERANCE > Math.abs(bestTarget.getYaw()));
-        } else {
-            // We don't have a target, so we're not lined up
-            return false;
+        List<List<TargetCorner>> ret = new ArrayList<>();
+        List<PhotonTrackedTarget> targets = pipelineResult.getTargets();
+
+        for (PhotonTrackedTarget target : targets) {
+            ret.add(target.getCorners());
         }
+
+        return ret;
+    }
+
+    private List<TargetCorner> parsedTargetCorners() {
+        List<List<TargetCorner>> cornerData = getTargetCorners();
+        if (cornerData == null) return null;
+
+        List<TargetCorner> ret = new ArrayList<>();
+
+        // assumes target contours are rectangles
+        for (List<TargetCorner> corners : cornerData) {
+            TargetCorner p1 = corners.get(0);
+            TargetCorner p2 = corners.get(1);
+            TargetCorner p3 = corners.get(2);
+            TargetCorner p4 = corners.get(3);
+
+            // following canvas axises (down is larger)
+            double midY = (p1.y + p2.y + p3.y + p4.y) / 4;
+
+            if (p1.y < midY) ret.add(p1);
+            if (p2.y < midY) ret.add(p2);
+            if (p3.y < midY) ret.add(p3);
+            if (p4.y < midY) ret.add(p4);
+        }
+
+        return ret;
+    }
+
+    private List<GlobalPoint> calculateGlobalPoints() {
+        List<TargetCorner> targetCorners = parsedTargetCorners();
+        if (targetCorners == null) return null;
+
+        List<GlobalPoint> ret = new ArrayList<>();
+
+        for (TargetCorner corner : targetCorners) {
+            // + 0.5 for 1 unit pixel plane
+            double nx = (1 / resolutionX) * (corner.x - resolutionX + 0.5);
+            double ny = (1 / resolutionY) * (resolutionY - corner.y + 0.5);
+            // coordinates on imaginary view plane
+            double x = vpw / 2 * nx;
+            double y = vph / 2 * ny;
+
+            pitch = Math.atan2(1, x);
+            yaw = Math.atan2(1, y);
```

Comment:
So you've calculated your pitch and your yaw, but you're missing the final step. you need your distance, which can be calculated in the post information here: https://www.chiefdelphi.com/t/what-does-limelight-skew-actually-measure/381167/7

Pay close attention to the vector stuff and having to account to camera pitch. Add both of those in if possible. Otherwise, this looks like good stuff,


It looks like I was initially wrong with the distance calculation as I was taking a heuristic approach to the problem: if the target isn't directly in front of us, the distance calculation will not be accurate. The linked post helps solve that, I think you understand vectors better than I do.